### DISCUSSION TITLE
[Filters] how to capture locals in the ERB filter?

### DISCUSSION THREAD

**Discussion by 0x1eef** [nanoc/nanoc#1611] [Get help] [answered]
Hello !

I have a Rules file that is sort of similar to this example:

```ruby
foo = [1,2,3,4]
bar = [5,6,7,8]

compile "/baz.html" do
  foo.each do |x|
    bar.each do |y|
      filter :erb, locals: {x: x, y: y}
      write "/#{x}/#{y}.html"
    end
  end
end
```

The problem is that the locals don't change past the first iteration (1, 5) in every file that is written. Is there a way around this ? Thanks!

**Comment by 0x1eef** [comment]
I tried a few things but in the end the best I could do is this:

```ruby
postprocess do 
  foo = [1,2,3,4]
  bar = [5,6,7,8]
  foo.each do |x|
    bar.each do |y|
      FileUtils.mkdir_p(File.join("output", x))
      File.write "output/#{x}/#{y}.html", 
                 ERB.new(File.read("./content/baz.html")).result_with_hash(x: x, y: y)
    end
end
```

**Comment by denisdefreyne** [comment, marked as answer]
Hey there!

Your original code is equivalent to the following:

```ruby
compile "/baz.html" do
  filter :erb, locals: {x: 1, y: 5}
  write "/1/5.html"

  filter :erb, locals: {x: 1, y: 6}
  write "/1/6.html"

  filter :erb, locals: {x: 1, y: 7}
  write "/1/7.html"

  # […]
end
```

After the first `:erb` filter, the ERB tags (`<%= … %>`) are already evaluated, so by the time the second `:erb` filter is run, there are no more ERB tags to evaluate.

I believe that you might look for something like this instead:

```ruby
foo.each do |x|
  bar.each do |y|
    compile "/baz.html", rep: "#{x}_#{y}" do
      filter :erb, locals: {x: x, y: y}
      write "/#{x}/#{y}.html"
    end
  end
end
```

This uses the “item representations” (or reps) feature of Nanoc to compile the same item, from scratch, multiple times. Note the `rep: "#{x}_#{y}`. This feature allows one input item in `content/` to have multiple files in `output/`.

See also:

* [Item representations](https://nanoc.app/doc/items-and-layouts/#item-representations)
* [Handling non-default representations](https://nanoc.app/doc/rules/#handling-non-default-representations)

**Comment by 0x1eef** [comment]
Thanks !! :heart: Works perfectly!
